engine/src/org/pentaho/di/repository/kdr/delegates/KettleDatabaseRepositoryConnectionDelegate.java

```diff
@@ -1917,6 +1897,15 @@ public synchronized void insertTableRow( String tablename, RowMetaAndData values
     }
   }
 
+  private synchronized PreparedStatement getPreparedStatement( String sql ) throws KettleDatabaseException {
```

Comment:
Even though there’s no time now to completely rewrite this tangled mass of code, we can’t just add another issue to an already bad problem. It has to start somewhere, and this is where. Please re-work your fix as follows:
•	Change sqlMap from a HashMap to a ConcurrentHashMap
•	Remove the synchronized keyword from your new method.
•	You should use the putIfAbsent method to add to the map